### ISSUE TITLE
Possible Deadlock on Connection Pool

### ISSUE THREAD

**Issue by marcusway** [nsidnev/fastapi-realworld-example-app#64]
This is less an issue and more a gotcha that I ran into when load testing, but I thought it would be worth noting since it took me a good bit of time to figure out, and I'm curious what people's thoughts on this might be.

If you're acquiring pool connections in `Depends` arguments, you can run into a connection leak that deadlocks your application if you exhaust the pool. This is a result of how dependencies are fulfilled in FastAPI, but following a pattern used in this example app can cause the behavior:

The `get_repository` dependency ([here](https://github.com/nsidnev/fastapi-realworld-example-app/blob/7f69c2a047b83fe3426fb15d57a53d636c9beb74/app/api/dependencies/database.py#L18)) returns an async generator, which gets you a `Repository` and handles the acquiring and release of the associated connection via the `async with` deal. The problem arises if you have multiple dependencies that acquire connections from the pool _and_ you exhaust the pool. Because the dependencies are executed in an `AsyncExitStack`, connections are released into the pool in the _reverse_ order that they acquired. 

For instance, the `update_current_user` [route](https://github.com/nsidnev/fastapi-realworld-example-app/blob/7f69c2a047b83fe3426fb15d57a53d636c9beb74/app/api/routes/users.py#L34)
```python
async def update_current_user(
    user_update: UserInUpdate = Body(..., embed=True, alias="user"),
    current_user: User = Depends(get_current_user_authorizer()),
    users_repo: UsersRepository = Depends(get_repository(UsersRepository)),
)
```

 depends on `get_repository(UsersRepository)` and `get_current_user_authorizer`, which transitively depends on a `get_repository(UsersRepository)` as well. When a request comes in, ideally you'll

1. acquire connection 1 in order to create the `UsersRepository` needed for `get_current_user_authorizer`
2. acquire connection 2 to create the second `UsersRepository` needed for `users_repo`
3. execute the method, hooray
3. release connection 2
4. release connection 1

But if you have a max_size of 1 on your connection pool, you'll

1. acquire connection 1
2. wait forever to acquire a connection because the release of connection 1 depends on the acquiring and release of this hypothetical connection 2, which depends on connection 1 being released

You'd probably never be so reckless as to have such a measly connection pool as that, but if you're under high load, what would have been brief pool exhaustion causing slight delays in acquiring connections quickly escalates into deadlock forever. Even if you add timeouts for acquiring connections and for query execution, you still suffer a pretty horrible performance hit this way, getting lots of timeout errors you wouldn't be getting otherwise, and holding connections open for a lot longer than you need to. It seems to me like the easiest way around this would be to have the `Repository` classes be long lived and take a `Pool` init arg rather than `Connection` so that they can handle the connection lifecycles themselves, but I'm curious if there are better ways, since `FastAPI` and a lot of this `async` business are pretty new to me.

I should note that this repo has been a tremendous resource, and I'm very grateful for its existence.

**Comment by nsidnev** [comment]
Hi! 

Yeah, I recently fixed this in #62 after it was first raised in #51. 

> take a `Pool` init arg rather than `Connection` so that they can handle the connection lifecycles themselves

Using a pool instead of a connection may be a way to fix this, but the asyncpg's pool does not offer a way to manage the transactions (which is actually understandable, since the transaction is tied to the connection, not to the pool) that are used in this application, so this will require changes to repositories that will try to get connections inside functions with transactions.

Instead, I fixed it, by providing a connection from pool per request as a separate dependency. This dependency will then be cached by FastAPI and deadlock should go away. This still limits an application that there can be no more than a **certain** number of clients, so if there are 11 clients at a time, then with the current pool(with 10 simultaneously open connections) the 11th client will wait for some of the other clients before it can receive a response. For this application, this is not critical, since all routes existing in this app are short-lived CRUD routes. but if websockets were used in this application, for example, then this approach would have to be changed to using a pool in repositories and acquiring connections where it need a new transaction.

> I should note that this repo has been a tremendous resource, and I'm very grateful for its existence.

Thanks! :smile:

**Comment by marcusway** [comment]
I'm late to the party! Thanks for the quick reply; that all makes sense to me.

> Instead, I fixed it, by providing a connection from pool per request as a separate dependency. This dependency will then be cached by FastAPI and deadlock should go away

Smaht.
